accounts-db/src/read_only_accounts_cache.rs

```diff
@@ -330,27 +325,62 @@ impl ReadOnlyAccountsCache {
             .expect("spawn accounts read cache evictor thread")
     }
 
-    /// Evicts entries until the cache's size is <= `target_data_size`
+    /// Evicts entries until the cache's size is <= `target_data_size`,
+    /// following the sampled LRU eviction method, where a sample of size
+    /// `evict_sample_size` is randomly selected from the cache, using the
+    /// provided `rng`.
     ///
-    /// Oldest entries are evicted first.
     /// Returns the number of entries evicted.
-    fn evict(
+    fn evict<R>(
         target_data_size: usize,
         data_size: &AtomicUsize,
+        evict_sample_size: usize,
         cache: &DashMap<ReadOnlyCacheKey, ReadOnlyAccountCacheEntry, AHashRandomState>,
-        queue: &Mutex<IndexList<ReadOnlyCacheKey>>,
-    ) -> u64 {
-        let mut num_evicts = 0;
+        rng: &mut R,
+        #[cfg(test)] mut callback: impl FnMut(&Pubkey, ReadOnlyAccountCacheEntry),
+    ) -> u64
+    where
+        R: Rng,
+    {
+        let mut num_evicts: u64 = 0;
         while data_size.load(Ordering::Relaxed) > target_data_size {
-            let Some(&key) = queue.lock().unwrap().get_first() else {
-                // if there are no more entries, we're done
-                break;
-            };
-            Self::do_remove(&key, cache, queue, data_size);
-            num_evicts += 1;
+            let mut key_to_evict = None;
+            let mut min_update_time = u64::MAX;
+            let mut remaining_samples = evict_sample_size;
+            while remaining_samples > 0 {
```

Comment:
```suggestion
            // NOTE: this can loop indefinitely if the cache is misconfigured and when we get here there aren't
            // at least `evict_sample_size` elements. We could break the loop on `cache.is_empty()` but calling
            // `is_empty()` and `len()` on a dashmap is very expensive as it requires iterating and locking all the
            // shards. So, avoid paying that cost and assume that when eviction triggers the cache contains
            // enough items. 
            while remaining_samples > 0 {
```

Reply:
Done